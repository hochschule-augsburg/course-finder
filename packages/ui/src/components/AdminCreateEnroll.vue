<script setup lang="ts">
import { trpc } from '@/api/trpc'
import { ref } from 'vue'

interface offeredCourseData {
  appointments: {
    dates: Array<{ from: Date; to: Date }>
    type: 'block' | 'irregular' | 'weekly'
  }
  extraInfo: string
  for: string[]
  maxParticipants: number
  minParticipants: number
  moduleCode: string
  moodleCourse: string
}

const sharedObject = ref<offeredCourseData[]>([])

const formData = ref({
  description: { de: '', en: '' }, // [I18n]
  end: '',
  start: '',
  title: { de: '', en: '' }, // [I18n]
})

async function createEnrollment() {
  const enrollmentPhaseData = {
    description: {
      de: formData.value.description.de,
      en: formData.value.description.en,
    },
    end: formData.value.end,
    offeredCourses: sharedObject.value,
    start: formData.value.start,
    title: { de: formData.value.title.de, en: formData.value.title.en },
  }
  try {
    await trpc.admin.enroll.phase.create.mutate(enrollmentPhaseData)
    console.log('Success creating enroll phase')
  } catch (e) {
    console.log('Error creating enroll phase')
  }
  sharedObject.value = []
}
</script>

<template>
  <VForm>
    <VContainer>
      <VRow justify="center">
        <VCol cols="12" sm="5"><h1>Create enrollment</h1></VCol>
        <VCol cols="12" sm="5"><VSpacer /></VCol>
      </VRow>
      <VRow justify="center">
        <VCol cols="12" sm="5">
          <VTextField
            v-model="formData.start"
            label="Start date"
            type="datetime-local"
            required
          />
        </VCol>
        <VCol cols="12" sm="5">
          <VTextField
            v-model="formData.end"
            label="End date"
            type="datetime-local"
            required
          />
        </VCol>
      </VRow>
      <VRow justify="center">
        <VCol cols="12" sm="5">
          <VTextField v-model="formData.title.en" label="Title (en)" required />
        </VCol>
        <VCol cols="12" sm="5">
          <VTextField v-model="formData.title.de" label="Title (de)" required />
        </VCol>
      </VRow>
      <VRow justify="center">
        <VCol cols="12" sm="10">
          <VTextarea
            v-model="formData.description.en"
            label="Description (en)"
            required
          />
        </VCol>
        <VCol cols="12" sm="10">
          <VTextarea
            v-model="formData.description.de"
            label="Description (de)"
            required
          />
        </VCol>
      </VRow>
      <VRow justify="center">
        <VCol cols="12" sm="5"><h1>Offered Courses</h1></VCol>
        <VCol cols="12" sm="5"><VSpacer /></VCol>
      </VRow>
      <VRow justify="center">
        <VCol cols="10">
          <OfferedCourses :offered-courses-array="sharedObject" />
        </VCol>
      </VRow>
      <VRow justify="center">
        <VCol cols="12" sm="5">
          <VBtn
            justify="center"
            text="Create enrollment"
            @click="createEnrollment"
          />
        </VCol>
        <VCol cols="12" sm="5"><VSpacer /></VCol>
      </VRow>
    </VContainer>
  </VForm>
</template>
