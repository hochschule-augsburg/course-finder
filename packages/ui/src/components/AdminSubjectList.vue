<script setup lang="ts">
import type { Course } from '@/stores/admin/AdminCoursesStore'

import { trpc } from '@/api/trpc'
import { useAdminCoursesStore } from '@/stores/admin/AdminCoursesStore'
import { merge } from 'lodash-es'
import { ref } from 'vue'

const adminStore = useAdminCoursesStore()

const showModalForm = ref(false)

const selectedSubject = ref<Course>(adminStore.courses[0])
const formData = ref({
  creditPoints: 0,
  editorUsername: '',
  extraInfo: '',
  facultyName: '',
  lecturers: '',
  semesterHours: 0,
  title: { de: '', en: '' }, // [I18n]
  varyingCP: '',
})

function selectSubject(s: Course) {
  selectedSubject.value = s
  formData.value.creditPoints = s.creditPoints
  formData.value.semesterHours = s.semesterHours
  formData.value.facultyName = s.facultyName ?? ''
  formData.value.title.de = s.title.de ?? ''
  formData.value.title.en = s.title.en ?? ''
  formData.value.lecturers = s.lecturers.join(', ')
  formData.value.varyingCP = varyingCPToString(s.varyingCP)
  formData.value.extraInfo = s.extraInfo ?? ''
  formData.value.editorUsername = s.editorUsername ?? ''

  showModalForm.value = true
}

function varyingCPToString(varyingCP: unknown) {
  if (
    typeof varyingCP === 'object' &&
    varyingCP !== null &&
    !Array.isArray(varyingCP)
  ) {
    const keyValuePairs = Object.entries(varyingCP).map(([key, value]) => {
      return `${key}: ${value}`
    })

    const resultString = keyValuePairs.join(', ')
    return resultString
  }
  return ''
}

function parseVaryingCP(input: string) {
  const dictionary: { [key: string]: number } = {}
  const pairs = input
    .split(',')
    .map((pair) => pair.trim())
    .filter((item) => item !== '')

  pairs.forEach((pair) => {
    const [key, value] = pair.split(':').map((item) => item.trim())
    dictionary[key] = Number(value)
  })
  return dictionary
}

async function saveSubject() {
  const inputData = {
    creditPoints: Number(formData.value.creditPoints),
    editorUsername: formData.value.editorUsername,
    extraInfo: formData.value.extraInfo,
    facultyName: formData.value.facultyName,
    lecturers: formData.value.lecturers
      .split(',')
      .map((item) => item.trim())
      .filter((item) => item !== ''),
    moduleCode: selectedSubject.value.moduleCode,
    semesterHours: Number(formData.value.semesterHours),
    title: {
      de: formData.value.title.de,
      en: formData.value.title.en,
    },
    varyingCP: parseVaryingCP(formData.value.varyingCP),
  }
  try {
    const result = await trpc.admin.courses.update.mutate(inputData)
    merge(selectedSubject.value, result)
  } catch (e) {
    console.log('Error updating Subject')
  }
  showModalForm.value = false
}
</script>

<template>
  <div>
    <VTable>
      <thead>
        <tr>
          <th class="text-left">Nr.</th>
          <th class="text-left">Name</th>
          <th class="text-left">Professor</th>
          <th>CP</th>
          <th>SWS</th>
          <th class="text-left">Edit</th>
        </tr>
      </thead>
      <tbody>
        <tr v-for="subject in adminStore.courses" :key="subject.moduleCode">
          <td>{{ subject.moduleCode }}</td>
          <td>{{ subject.title.en }}</td>
          <td>{{ subject.lecturers.toString() }}</td>
          <td>{{ subject.creditPoints }}</td>
          <td>{{ subject.semesterHours }}</td>
          <td>
            <VBtn @click="selectSubject(subject)">
              <VIcon size="25">mdi-pencil</VIcon>
            </VBtn>
          </td>
        </tr>
      </tbody>
      <VBtn>Save</VBtn>
    </VTable>
    <VDialog
      v-model="showModalForm"
      min-width="800"
      transition="false"
      width="auto"
    >
      <VCard
        :title="`Edit - ${selectedSubject.title.en}`"
        prepend-icon="mdi-pencil"
      >
        <VCardText>
          <VRow dense>
            <VCol cols="12" md="4" sm="6">
              <VTextField
                v-model="formData.title.en"
                label="Title (en)"
                required
              />
            </VCol>
            <VCol cols="12" md="4" sm="6">
              <VTextField
                v-model="formData.title.de"
                label="Title (de)"
                required
              />
            </VCol>
            <VCol>
              <VTextField
                v-model="formData.editorUsername"
                label="Editor username"
                required
              />
            </VCol>
            <VCol cols="12" md="4" sm="6">
              <VTextField
                v-model="formData.semesterHours"
                label="Semester Hours (SWS)"
                type="number"
                required
              />
            </VCol>
            <VCol cols="12" md="4" sm="6">
              <VTextField
                v-model="formData.creditPoints"
                label="Credit points (CP)"
                type="number"
                required
              />
            </VCol>
            <VCol>
              <VTextField
                v-model="formData.varyingCP"
                hint="E.g. IN: 5, WIN:8,"
                label="Varying CP"
                required
              />
            </VCol>
            <VCol cols="12">
              <VTextField
                v-model="formData.lecturers"
                hint="separate with comma"
                label="Lecturers*"
                required
              />
            </VCol>
            <VCol>
              <VTextarea
                v-model="formData.extraInfo"
                label="Extra information"
                required
              />
            </VCol>
          </VRow>
          <small class="text-caption text-medium-emphasis"
            >*separate multiple elements with comma</small
          >
        </VCardText>
        <VDivider />
        <template #actions>
          <VBtn text="Cancel" @click="showModalForm = false" />
          <VBtn class="ms-auto" text="Save" @click="saveSubject" />
        </template>
      </VCard>
    </VDialog>
  </div>
</template>
