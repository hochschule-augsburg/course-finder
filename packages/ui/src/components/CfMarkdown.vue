<script lang="ts" setup>
import type { Options, PluginSimple } from 'markdown-it'

import VueMarkdown from 'vue-markdown-render'

const props = defineProps<{
  source?: string
}>()
const newTabLinkPlugin: PluginSimple = (md) => {
  const defaultRender =
    md.renderer.rules.link_open ||
    function (tokens, idx, options, env, self) {
      return self.renderToken(tokens, idx, options)
    }

  md.renderer.rules.link_open = function (tokens, idx, options, env, self) {
    tokens[idx].attrSet('target', '_blank')
    tokens[idx].attrJoin('onclick', 'event.stopPropagation()')

    return defaultRender(tokens, idx, options, env, self)
  }
}

const options: Options = {
  linkify: true,
}
const plugins = [newTabLinkPlugin]
</script>

<template>
  <VueMarkdown
    :options="options"
    :plugins="plugins"
    :source="props.source"
    class="cf-markdown-component"
  />
</template>

<!-- eslint-disable-next-line vue/enforce-style-attribute -->
<style lang="scss">
.cf-markdown-component {
  ul {
    padding: revert;
  }
  ol {
    padding: revert;
  }
}
</style>
