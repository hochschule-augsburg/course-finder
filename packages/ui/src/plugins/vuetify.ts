/**
 * plugins/vuetify.ts
 *
 * Framework documentation: https://vuetifyjs.com`
 */

// Styles
import '@mdi/font/css/materialdesignicons.css'
// Composables
import { createVuetify } from 'vuetify'
import 'vuetify/styles'

// https://vuetifyjs.com/en/introduction/why-vuetify/#feature-guides
export default createVuetify({
  theme: {
    defaultTheme: 'dark',
  },
})
