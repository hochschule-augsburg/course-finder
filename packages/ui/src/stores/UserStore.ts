import { trpc } from '@/api/trpc'
import { defineStore } from 'pinia'
import { ref } from 'vue'

import { UserExtended } from '../../../api/src/libExports'

export const useUserStore = defineStore('user', () => {
  const user = ref<UserExtended>()
  void init()
  return { login, logout, user }
  async function login(username: string, password: string, otp?: string) {
    if (otp) {
      return twoFA(username, otp)
    }
    const result = await trpc.auth.login.mutate({
      password,
      username,
    })
    if (typeof result === 'object') {
      user.value = result
    }
    return result
  }
  async function twoFA(username: string, otp: string) {
    const result = await trpc.auth.twoFA.mutate({
      otp,
      username,
    })
    if (typeof result === 'object') {
      user.value = result
    }
    return result
  }
  async function logout() {
    await trpc.auth.logout.mutate()
    user.value = undefined
  }
  async function init() {
    user.value = await trpc.auth.getUser.query()
  }
})
