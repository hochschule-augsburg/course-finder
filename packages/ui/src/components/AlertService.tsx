import { defineComponent, ref } from 'vue'
import { VAlert, VOverlay } from 'vuetify/components'

type AlertProps = InstanceType<typeof VAlert>['$props']

export type AlertOptions = { duration?: number } & Pick<
  AlertProps,
  'color' | 'text' | 'title' | 'width'
>

let alert: { show: (options: AlertOptions) => void } | undefined

export const alertService = {
  show(options: AlertOptions) {
    alert?.show(options)
  },
}

export const ServiceAlert = defineComponent(
  () => {
    const alertOptions = ref<AlertOptions>()
    const show = ref(false)
    alert = {
      show: (options) => {
        alertOptions.value = options
        show.value = true
        setTimeout(() => {
          show.value = false
        }, options.duration ?? 1000)
      },
    }

    if (import.meta.hot) {
      import.meta.hot.invalidate()
    }
    return () => (
      <VOverlay
        activator="parent"
        modelValue={show.value}
        origin="center center"
        scrim={false}
        style={{
          alignItems: 'end',
          justifyContent: 'left',
          padding: '0.5rem',
        }}
        zIndex={3000}
      >
        <VAlert {...alertOptions.value} maxWidth={500} />
      </VOverlay>
    )
  },
  { name: 'ServiceAlert' },
)
