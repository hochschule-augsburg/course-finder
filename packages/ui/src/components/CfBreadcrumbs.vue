<script lang="ts" setup>
import { mdiChevronRight } from '@mdi/js'
import { computed } from 'vue'
import { useI18n } from 'vue-i18n'
import { useRoute, useRouter } from 'vue-router'
import { VBreadcrumbs, VBreadcrumbsItem, VIcon } from 'vuetify/components'

const { t } = useI18n()
const route = useRoute()
const router = useRouter()

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const routeItems = computed<any>(() => {
  if (route.path === '/') {
    return []
  }

  const breadcrumbItems: {
    disabled: boolean;
    href: string;
    noBreadcrumbs: boolean;
    params: Record<string, string>;
    text: string;
  }[] = [
    {
      disabled: false,
      href: '/',
      noBreadcrumbs: false,
      params: {},
      text: '.index',
    },
  ]

  for (const [i, match] of route.matched.slice(0, -1).entries()) {
    if (match.meta.noBreadcrumbs) {
      continue
    }
    const text = match.path.split('/').join('.')
    breadcrumbItems.push({
      disabled: i === route.matched.length - 2, // disable last
      href: match.path,
      noBreadcrumbs: match.meta.noBreadcrumbs,
      params: route.params,
      text: match.children.length ? `${text}.index` : text,
    })
  }

  return breadcrumbItems
})
</script>

<template>
  <VBreadcrumbs :items="routeItems" class="ml-9">
    <template #divider>
      <VIcon :icon="mdiChevronRight" />
    </template>
    <template #item="{ item }">
      <VBreadcrumbsItem :disabled="item.disabled" :to="item.href">
        {{
          // @ts-expect-error wrong typing in vuetify
          t(`pages${item.text}`, item.params)
        }}
      </VBreadcrumbsItem>
    </template>
  </VBreadcrumbs>
</template>
