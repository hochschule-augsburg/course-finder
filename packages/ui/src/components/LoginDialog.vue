<script setup lang="ts">
import { computed, ref } from 'vue'
import { useI18n } from 'vue-i18n'
import {
  VBtn,
  VCard,
  VCardActions,
  VCardText,
  VCardTitle,
  VForm,
  VSpacer,
  VTextField,
} from 'vuetify/components'

import { useUserStore } from '@/stores/UserStore'

const emit = defineEmits<{
  success: []
}>()

const { t } = useI18n()

const userStore = useUserStore()

const username = ref('')
const password = ref('')
const otp = ref('')

const isValid = computed(() => !!(username.value && password.value))
const twoFANeeded = ref(false)
const pending = ref(false)
const error = ref<string>()

async function login() {
  pending.value = true
  let result: 'unknown-error' | Awaited<ReturnType<typeof userStore.login>>
  try {
    result = await userStore.login(username.value, password.value)
  } catch (e) {
    console.error(e)
    error.value = 'global.unknown-error'
    return
  } finally {
    pending.value = false
  }
  if (typeof result === 'object') {
    error.value = undefined
    emit('success')
    return
  }
  if (result === 'two-fa-required') {
    twoFANeeded.value = true
    return
  }
  error.value = `error.${result}`
}

async function twoFALogin() {
  pending.value = true
  let result: 'unknown-error' | Awaited<ReturnType<typeof userStore.login>>
  try {
    result = await userStore.login(username.value, password.value, otp.value)
  } catch (e) {
    console.error(e)
    error.value = 'global.unknown-error'
    return
  } finally {
    pending.value = false
  }
  if (typeof result === 'object') {
    emit('success')
    return
  }
  error.value = `error.two-fa.${result}`
}

function validUsername(input: string) {
  return (
    !/[*().&[\]`|%^?{}! ,\\#+<>;"=']/.test(input) ||
    t('error.invalid-characters')
  )
}
</script>

<template>
  <VCard class="login-dialog-host">
    <VCardTitle>Login</VCardTitle>
    <VCardText>
      <VForm v-if="!twoFANeeded" @submit="login">
        <VTextField
          v-model="username"
          :error="!!error"
          :label="t('username')"
          :rules="[validUsername]"
          autofocus
          required
          @update:model-value="error = undefined"
        />
        <VTextField
          v-model="password"
          :error-messages="error ? t(error) : undefined"
          :label="t('password')"
          type="password"
          required
          @keyup.enter="login"
          @update:model-value="error = undefined"
        />
      </VForm>
      <VTextField
        v-if="twoFANeeded"
        v-model="otp"
        :label="t('two-fa-code')"
        type="number"
        hide-spin-buttons
        @keyup.enter="twoFALogin"
      />
    </VCardText>
    <VCardActions>
      <VSpacer />
      <VBtn
        :disabled="!isValid"
        :loading="pending"
        color="primary"
        variant="tonal"
        @click="!twoFANeeded ? login() : twoFALogin()"
      >
        {{ t('login') }}
      </VBtn>
    </VCardActions>
  </VCard>
</template>

<style scoped lang="scss">
.login-dialog-host {
  min-width: 20rem;
}
</style>

<i18n lang="yaml">
en:
  login: Login
  username: Username
  password: Password
  two-fa-code: Two-factor code
  error:
    service-not-available: Service not available
    two-fa-required: Two-factor authentication required
    invalid-credentials: Invalid credentials
    already-logged-in: Already logged in
    invalid-characters: Invalid characters
    tow-fa:
      code-invalid: Invalid code
      code-expired: Code expired
de:
  login: Einloggen
  username: Benutzername
  password: Passwort
  two-fa-code: Zwei-Faktor-Code
  error:
    service-not-available: Dienst nicht verf端gbar
    two-fa-required: Zwei-Faktor-Authentifizierung erforderlich
    invalid-credentials: Ung端ltige Anmeldeinformationen
    already-logged-in: Bereits angemeldet
    invalid-characters: Ung端ltige Zeichen
    tow-fa:
      code-invalid: Ung端ltiger Code
      code-expired: Code abgelaufen
</i18n>
