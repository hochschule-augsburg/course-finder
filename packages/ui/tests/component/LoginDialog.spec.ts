/* eslint-disable @typescript-eslint/no-explicit-any */

import LoginDialog from '@/components/LoginDialog.vue'
import { useUserStore } from '@/stores/UserStore'
import { createTestingPinia } from '@pinia/testing'
import { flushPromises, mount } from '@vue/test-utils'
import { createVuetify } from 'vuetify'
import { VBtn, VTextField } from 'vuetify/components'

import { i18nForTests } from '../test-utils/i18nForTests'

vi.mock('@/stores/UserStore')

describe('LoginDialog.vue', () => {
  it('emits success event on successful login', async () => {
    //@ts-ignore
    vi.mocked(useUserStore).mockReturnValue({
      login: () => ({}) as any,
    })
    const wrapper = mountComponent()

    wrapper.findComponent(VBtn).vm.$emit('click')
    await flushPromises()

    expect(wrapper.emitted('success')).toBeTruthy()
  })

  it('sets error value on failed login', async () => {
    //@ts-ignore
    vi.mocked(useUserStore).mockReturnValue({
      login: () => Promise.reject(new Error('bla')),
    })
    const wrapper = mountComponent()

    wrapper.findComponent(VBtn).vm.$emit('click')
    await flushPromises()

    expect(
      wrapper
        .findAllComponents(VTextField)
        .find((e) => e.props('label') === 'Password')
        ?.props('errorMessages'),
    ).toBe('global.unknown-error')
  })

  it('should do twoFA', async () => {
    const username = 'testuser'
    const password = 'testpassword'
    const otp = '123456'
    let allowLogin = false

    const loginMock = vi.fn(() => {
      if (allowLogin) {
        return {} as any
      }
      return Promise.resolve('two-fa-required' as const)
    })
    //@ts-ignore
    vi.mocked(useUserStore).mockReturnValue({
      login: loginMock,
    })
    const wrapper = mountComponent()

    wrapper
      .findAllComponents(VTextField)
      .find((e) => e.props('label') === 'Username')
      ?.vm.$emit('update:modelValue', username)
    wrapper
      .findAllComponents(VTextField)
      .find((e) => e.props('label') === 'Password')
      ?.vm.$emit('update:modelValue', password)
    await flushPromises()

    wrapper.findComponent(VBtn).vm.$emit('click')
    await flushPromises()

    expect(loginMock).toHaveBeenCalledWith(username, password)

    allowLogin = true

    wrapper
      .findAllComponents(VTextField)
      .find((e) => e.props('label') === 'Two-factor code')
      ?.vm.$emit('update:modelValue', otp)
    await flushPromises()
    wrapper.findComponent(VBtn).vm.$emit('click')
    await flushPromises()

    expect(loginMock).toHaveBeenCalledWith(username, password, otp)
    expect(wrapper.emitted('success')).toBeTruthy()
  })

  it('sets error value on failed twoFALogin', async () => {
    const errorText = 'code-expired'
    let loginReturn: Promise<typeof errorText> = Promise.resolve(errorText)
    const loginMock = vi.fn(() => loginReturn)
    //@ts-ignore
    vi.mocked(useUserStore).mockReturnValue({
      login: loginMock,
    })
    const wrapper = mountComponent({ shallow: true })

    // @ts-ignore
    await wrapper.vm.twoFALogin()

    // @ts-ignore
    expect(wrapper.vm.error).toBe(`error.two-fa.${errorText}`)

    loginReturn = Promise.reject()

    // @ts-ignore
    await wrapper.vm.twoFALogin()

    // @ts-ignore
    expect(wrapper.vm.error).toBe('global.unknown-error')
  })
})

function mountComponent(opts: { shallow?: boolean } = {}) {
  return mount(LoginDialog, {
    global: {
      plugins: [createTestingPinia(), i18nForTests(), createVuetify()],
    },
    shallow: opts.shallow,
  })
}
