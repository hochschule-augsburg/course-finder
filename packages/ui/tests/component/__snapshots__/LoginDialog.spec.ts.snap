// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginDialog.vue > match snapshot 1`] = `
"<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background" style="width: 100%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-title">Login</div>
  <div class="v-card-text">
    <form class="v-form" novalidate="">
      <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-locale--is-ltr v-text-field">
        <!---->
        <div class="v-input__control">
          <div class="v-field v-field--center-affix v-field--variant-filled v-theme--light v-locale--is-ltr">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background" style="width: 100%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-0">
                <!---->Username
              </label><label class="v-label v-field-label" for="input-0">
                <!---->Username
              </label>
              <!----><input size="1" type="text" id="input-0" aria-describedby="input-0-messages" required="" class="v-field__input">
              <!---->
            </div>
            <!---->
            <!---->
            <div class="v-field__outline">
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <div class="v-input__details">
          <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages" role="alert" aria-live="polite" id="input-0-messages">
            <!---->
          </transition-group-stub>
          <!---->
        </div>
      </div>
      <div class="v-input v-input--horizontal v-input--center-affix v-input--density-default v-locale--is-ltr v-text-field">
        <!---->
        <div class="v-input__control">
          <div class="v-field v-field--center-affix v-field--variant-filled v-theme--light v-locale--is-ltr">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px; left: 50%; transform: translateX(-50%);" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background" style="width: 100%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <!---->
            <div class="v-field__field" data-no-activator=""><label class="v-label v-field-label v-field-label--floating" aria-hidden="true" for="input-2">
                <!---->Password
              </label><label class="v-label v-field-label" for="input-2">
                <!---->Password
              </label>
              <!----><input size="1" type="password" id="input-2" aria-describedby="input-2-messages" required="" class="v-field__input">
              <!---->
            </div>
            <!---->
            <!---->
            <div class="v-field__outline">
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <div class="v-input__details">
          <transition-group-stub name="slide-y-transition" tag="div" appear="false" persisted="false" css="true" class="v-messages" role="alert" aria-live="polite" id="input-2-messages">
            <!---->
          </transition-group-stub>
          <!---->
        </div>
      </div>
    </form>
    <!--v-if-->
  </div>
  <div class="v-card-actions">
    <div class="v-spacer"></div><button type="button" class="v-btn v-btn--disabled v-btn--slim v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-text" disabled=""><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
      <!----><span class="v-btn__content" data-no-activator="">Login </span>
      <!---->
      <!---->
    </button>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
